plugins/woocommerce-blocks/tests/e2e/tests/filter-blocks/stock-status.block_theme.spec.ts

```diff
@@ -0,0 +1,88 @@
+/**
+ * External dependencies
+ */
+import { test, expect } from '@woocommerce/e2e-playwright-utils';
+import path from 'path';
+
+/**
+ * Internal dependencies
+ */
+import {
+	GeneratedPost,
+	createPostFromTemplate,
+} from '../../utils/create-dynamic-content';
+
+const TEMPLATE_PATH = path.join( __dirname, 'stock-status.handlebars' );
+
+test.describe( 'Product Filter: Stock Status Block', async () => {
+	let defaultBlockPost: GeneratedPost;
+	let dropdownBlockPost: GeneratedPost;
+
+	test.beforeAll( async () => {
+		defaultBlockPost = await createPostFromTemplate(
+			'Product Filter Stock Status Block',
+			TEMPLATE_PATH,
+			{}
+		);
+
+		dropdownBlockPost = await createPostFromTemplate(
+			'Product Filter Stock Status Block Dropdown',
+			TEMPLATE_PATH,
+			{
+				attributes: {
+					displayStyle: 'dropdown',
+				},
+			}
+		);
+	} );
+
+	test( 'By default it renders a checkbox list with the available stock statuses', async ( {
+		page,
+	} ) => {
+		await page.goto( defaultBlockPost.url );
+
+		const stockStatuses = page.locator(
+			'.wc-block-components-checkbox__label'
+		);
+
+		await expect( stockStatuses ).toHaveCount( 2 );
+		await expect( stockStatuses.nth( 0 ) ).toHaveText( 'In stock' );
+		await expect( stockStatuses.nth( 1 ) ).toHaveText( 'Out of stock' );
+	} );
+
+	test( 'Selecting a stock status filters the list of products', async ( {
+		page,
+	} ) => {
+		await page.goto( defaultBlockPost.url );
+
+		const outOfStockCheckbox = page.getByLabel( 'Out of stock' );
+		outOfStockCheckbox.click();
+
+		// wait for navigation
+		await page.waitForURL( /.*filter_stock_status=outofstock.*/ );
+
+		const products = page.locator( '.wc-block-product' );
+		const productCount = await products.count();
+
+		expect( productCount ).toBe( 1 );
+	} );
+
+	test( 'When displayStyle is dropdown, a dropdown is displayed with the available stock statuses', async ( {
+		page,
+	} ) => {
+		await page.goto( dropdownBlockPost.url );
+
+		const dropdownLocator = page.locator( '.wc-interactivity-dropdown' );
+
+		await expect( dropdownLocator ).toBeVisible();
+		await dropdownLocator.click();
+
+		await expect( page.getByText( 'In stock' ) ).toBeVisible();
+		await expect( page.getByText( 'Out of stock' ) ).toBeVisible();
+	} );
+
+	test.afterAll( async () => {
+		await defaultBlockPost.deletePost();
+		await dropdownBlockPost.deletePost();
+	} );
```

Comment:
Will running multiple (independent?) tests against the same post be safe, stability-wise? Generally, if we'd like to run a multi-step scenario test against the same post, we can use the [`test.step` API](https://playwright.dev/docs/api/class-test#test-step). Otherwise, using a fresh post for each `test` would generally be recommended to avoid running the test in a (potentially) non-deterministic environment. To achieve that without duplicating code, we could consider turning `createPostFromTemplate` into a [Playwright Fixture](https://playwright.dev/docs/test-fixtures), where we'd handle all the setup and teardown for each block post, saving precious LOCs. 😄 

Let me know if that makes sense!

Reply:
Yes great question and point. It depends! And its the same with the existing templates we use (which are created once before the test framework is run). 

If you interacted with a block in the editor then certainly it would impact how it would behave on the front-end. To me it feels like its the responsibility of the developer to determine whether what they're doing should be torn down and when or be aware if what they're doing would have a side effect.

> Otherwise, using a fresh post for each test would generally be recommended to avoid running the test in a (potentially) non-deterministic environment. To achieve that without duplicating code, we could consider turning createPostFromTemplate into a [Playwright Fixture](https://playwright.dev/docs/test-fixtures), where we'd handle all the setup and teardown for each block post, saving precious LOCs. 😄

This API looks awesome thank you, I didn't realize playwright has so many useful APIs 🙇 

I think we should balance between using this fixture API and approaches like I've used here based on performance (and whether the post has side effects) mainly for performance reasons. Because creating the post and tearing it down takes some time to do every single test. But I'm still keen to be convinced I'm wrong here, tbh I haven't done a lot of PlayWright testing so maybe I'm wrong.

Reply:
> Because creating the post and tearing it down takes some time to do every single test. But I'm still keen to be convinced I'm wrong here, tbh I haven't done a lot of PlayWright testing so maybe I'm wrong.

I don't think you're wrong - it will likely take more time than just calling `page.goto`, but it's the kind of a tradeoff that usually needs to be considered: Do we want to wait a little longer but provide better resilience or is it ok to risk some flakiness for the sake of speed? There are many factors to it, but from my experience, e2e flakiness has never been welcomed by fellow devs! 😅 

Isolating tests has been a good practice stability-wise, e.g., in the core Gutenberg testing. To be fair, full isolation is still impossible since we're testing against a single wp-env instance, where the editor preferences are persistent (changing a preference will bleed into parallel tests). Ideally, though, each test should be completely isolated.

Reply:
> Because creating the post and tearing it down takes some time to do every single test.

I think it won't be a big issue if we use REST API to manipulate posts.

Reply:
Resolved in https://github.com/woocommerce/woocommerce/pull/44223/commits/64b60b48aeacf4bc79d38f9133e1d05261e7344f